Fix rustls features

stormshield-gt opened the pull request:
I've gone too swiftly while developing https://github.com/fussybeaver/bollard/pull/441 and I didn't test all the cases.
This PR amend that, and add extra checks to the

fussybeaver commented:
Hey, thanks for this and sorry for the delay in review... 

I have a couple of questions.. 

None of the checks fail on master currently:
 - `cargo check --no-default-features --all-targets --features ssl`
 - `cargo check --no-default-features --all-targets --features ssl_providerless`
 - `cargo check --no-default-features --all-targets --features aws-lc-rs`

Is there a compile that failed on master that passes on your PR ? The PR otherwise looks fine.

Finally, if this really is an issue, then it looks like we may need to backport this as a separate patch release.

stormshield-gt commented:
Thanks for the review!

> None of the checks fail on master currently

Good catch, I've added a test that currently failed the master https://app.circleci.com/pipelines/github/fussybeaver/bollard/1395/workflows/d2053e50-d1ae-4019-b5c5-63d73350cf24/jobs/14227
and then pushed the fix.

To do so, I removed the `webpki` feature. Indeed, as stipulated in the [`webpki_roots` readme warning](https://github.com/rustls/webpki-roots?tab=readme-ov-file#warning), `webpki_roots` shouldn't be used in production except in some rare cases, as it requires recompiling each time the roots evolve.
I still use it for the test, as does `rustls` organization repositories.

> Finally, if this really is an issue, then it looks like we may need to backport this as a separate patch release.

If you agree with the `webpki` feature removal, then it would necessitate a `minor` release.

fussybeaver commented on src/docker.rs:
```diff
@@ -388,16 +388,19 @@ impl DockerClientCertResolver {
-            path: self.ssl_key.to_owned(),
-        })?;
+        let signing_key = CryptoProvider::get_default()
+            .expect("no process-level CryptoProvider available -- call CryptoProvider::install_default() before this point")
```
We probably shouldn't panic here, as we would need to document clearly that it can panic in the connection methods. Could you create a new error type in `errors.rs` with the appropriate message and return that ?

stormshield-gt replied:
The underlying reasoning was if the user failed to install a provider, this error is just unrecoverable and the code should just panic. 
That's something that crates inside the `rustls` organization do.
For instance, see those lines in [rustls](https://github.com/rustls/rustls/blob/e52853a08fa366bb6f3253e932f5608c8b580d39/rustls/src/crypto/mod.rs#L265) and [rustls-platform-verifier](https://github.com/rustls/rustls-platform-verifier/blob/1ba897603508c55a7c350266238b9e9f2c7a578a/rustls-platform-verifier/src/verification/mod.rs#L113)

Anyway I don't feel strongly about this and would be happy to change if you prefer the other way.

fussybeaver replied:
Ok. It's understandable perhaps to panic with incorrect compile time flags, but we don't quite know how a docker client is invoked (maybe in a parallel thread), would be best to return an error in any case.

fussybeaver commented on Cargo.toml:
```diff
@@ -36,7 +38,6 @@ json_data_content = []
 ssl = ["ssl_providerless", "rustls/ring"]
 aws-lc-rs = ["ssl_providerless", "rustls/aws-lc-rs"]
 ssl_providerless = ["home", "hyper-rustls", "rustls", "rustls-native-certs", "rustls-pemfile", "rustls-pki-types"]
-webpki = ["ssl", "dep:webpki-roots"]
```
Fair point about these being unrecommended for production, but let's keep these for the moment in case someone is actually using them.

stormshield-gt replied:
Sure, I re-add it. Maybe this should be removed during the next breaking release

fussybeaver replied:
Let's have a separate issue and changeset for this and communicate the change in the README.

fussybeaver commented:
So.. what I understand now, the problem currently in `master` is that we're still using the `ring` crypto implementation, even if the user specifies the `aws-lc-rs` crypto implementation and forgets to use `CryptoProvider::install_default()` ?

stormshield-gt commented:
The user can only use the `ring` provider feature, and the `aws-lc-rs` and `ssl-providerless` aren't usable because the necessary functions are gated under the feature `ssl` instead of `ssl-providerless`. 
So if someone wants to use the `aws-lc-rs` feature, it will have [this compiler error](https://app.circleci.com/pipelines/github/fussybeaver/bollard/1395/workflows/d2053e50-d1ae-4019-b5c5-63d73350cf24/jobs/14227).

stormshield-gt commented:
The current failed on CI is also on master with the command `cargo check --no-default-features --all-targets --features ssl`. As this wasn't checked in CI before, https://github.com/fussybeaver/bollard/pull/465 seems to have introduced the failure. 
Maybe a separate PR should be open to fix this ?

fussybeaver commented on .circleci/config.yml:
```diff
@@ -12,9 +12,12 @@ jobs:
+      - run: docker run -ti -e DOCKER_CERT_PATH=/certs -e DOCKER_HOST='tcp://test.example.com:2376' -e DOCKER_TLS_VERIFY=1 --volumes-from certs --rm --link test-docker-daemon:docker bollard cargo test --features test_ssl,ssl -- --test test_connect_with_defaults
+      - run: docker run -ti -e DOCKER_CERT_PATH=/certs -e DOCKER_HOST='tcp://test.example.com:2376' -e DOCKER_TLS_VERIFY=1 --volumes-from certs --rm --link test-docker-daemon:docker bollard cargo test --features test_ssl,aws-lc-rs -- --test test_connect_with_defaults
+      - run: docker run -ti -e DOCKER_CERT_PATH=/certs -e DOCKER_HOST='https://test.example.com:2376' --volumes-from certs --rm --link test-docker-daemon:docker bollard cargo test --features test_ssl,ssl -- --test test_connect_with_defaults
+      - run: docker run -ti -e DOCKER_CERT_PATH=/certs -e DOCKER_HOST='https://test.example.com:2376' --volumes-from certs --rm --link test-docker-daemon:docker bollard cargo test --features test_ssl,aws-lc-rs -- --test test_connect_with_defaults
```
I think you'll need to add feature `http` to your test cases, since https://github.com/fussybeaver/bollard/pull/465  removed I/O libraries from required features.

stormshield-gt replied:
Thanks for the suggestion, I added the `http` feature

fussybeaver commented:
Thanks for sticking with this! If you can document the entry functions to connecting with ssl that it can panic, then I'm happy to merge this PR.

stormshield-gt commented:
I've added a panic section to the documentation of the concerned functions

fussybeaver commented:
Thanks for putting the work into this
